Add a list of active forks.

player-03 opened the pull request:
People have been asking what the current status is, so I think it would be nice to add a message to the readme. This message would let people know that you're done making updates and that a couple of active forks exist. I included three features from each fork to help them decide, plus a disclaimer that you are only responsible for deepcake/echo.

player-03 committed:
Add a list of active forks.

player-03 commented:
@onehundredfeet, is my description of your repo ok, or should I put something else?

zommerfelds commented on README.md:
```diff
@@ -6,7 +6,13 @@ Initially created to learn the power of macros.
 
-#### Wip
+### Forks
+deepcake/echo is considered complete, but multiple forks are still under active development.
```
"is considered complete" sounds a bit ambiguous. How about "is no longer being actively maintained"?

player-03 replied:
Yeah, I guess that's more precise.

zommerfelds commented on README.md:
```diff
@@ -6,7 +6,13 @@ Initially created to learn the power of macros.
+ * [player-03/echoes](https://github.com/player-03/echoes) focuses on ease of use for medium-scale games. Features include entity templates, customizable timesteps, and system priorities.
+ * [onehundredfeet/hmecs](https://github.com/onehundredfeet/hmecs) focuses on performance at scale. Features include object pooling, customizable storage, and singleton components.
+
+Disclaimer: this list is not an endorsement, and deepcake can't guarantee that these forks are backwards-compatible. More details can be found in each fork's readme.
```
Seems unnecessary to add this. But up to you :)

player-03 replied:
Wasn't sure if deepcake would want something like that. They did mention backwards compatibility being a concern...

I'll remove it for now.

player-03 committed:
Update fork list based on feedback.

onehundredfeet commented:
Wow.  Thanks for putting this together. I hadn't really thought of the 'focus' as I really just added what I needed. But yes, I think you've covered the high level.

1) Memory Performance - Allowing different storage mechanisms to reduce waste
2) CPU Performance - Object pooling to reduce the demand on the GC
3) Network Architecture Support 
- World Segregation - Allowing the ECS to segment entities by 'worlds' instead of a single world in a trivial, non-memory intensive way
- Explicit updates - You can update systems lists manually

I use explicit system lists rather than letting the ecs prioritize it.  

Where I'm headed is:
4) Making multi-threading much easier to do in general, which is described in the readme.md

deepcake commented:
What a good initiative! I was just thinking about something like this. Will merge it in a few days.

player-03 commented:
> I use explicit system lists rather than letting the ecs prioritize it.

Same, but it's hard to condense [this](https://github.com/player-03/echoes#priority) down into two words. Is there a better way of phrasing it?

Edit; I don't want to repeat "customizable" in the same sentence if I don't have to. I could try "event priorities," since at least OpenFL programmers may be familiar with the concept. "update listener priorities" would be more specific, but maybe too specific for a broad-strokes overview.

onehundredfeet commented:
It's a system priority system.  I think you've captured it.  I just rely on explicit ordering instead of prioritization.  I break my systems into two if they need to run in two different spots.  

You may need to be careful with priorities on specific update functions and what is expected.  i.e. does a high priority update function get called before all other lower priorities functions on the current system?  or within the current system list? or globally?  It's not clear to me on reading your docs, which one it is.

player-03 commented:
> I just rely on explicit ordering instead of prioritization.

I tried that first, but couldn't figure out a way to handle complicated use cases. When you're writing a game on top of a game engine on top of Echoes, sometimes the game will need to insert a system before (or in the middle of) the game engine's system list.

My first thoughts were to let you set an index, or insert your system before a specific other system. But the index relies on the game engine having an exact number of systems (and what if it gets updated?). The "before" option is better, but what if someday I write a different game that cherry-picks systems and initializes them in a different order? Store which system goes before which, and you could end up with two systems each trying to go before the other.

> I break my systems into two if they need to run in two different spots.

But what if you write two interconnected systems and the user only adds one of them? Sure you could write a note "if you use this system, make sure to use the other as well," but notes are easy to miss. And when the code broke it might not be obvious why.

My code does actually use two systems under the hood, but `SystemList` makes sure they are added and removed in tandem.

I think explicit ordering is ideal for at least 90% of cases, but there are a few not-uncommon cases where priority works better.

> does a high priority update function get called before all other lower priorities functions on the current system? or within the current system list? or globally? It's not clear to me on reading your docs, which one it is.

This is what I was trying to explain in the part starting with "Note that `@:priority` is only used to sort the parent `SystemList`." I'll do another revision.

onehundredfeet commented:
Ah, I see.  My audience is the writer of the game, not the writer of middleware.  I can see your challenge now.

I think that your priority system solves a chunk of what you're looking for.  It just may be fairly easy for a person to make a mistake and break the system.

For example, if I put a clean-up function in several systems that's supposed to run last (low priority?), after all the systems have done their work, but the user then puts those systems into separate lists, that won't happen anymore.  You will get the first system running to completion and the second and then the third, etc.  The user still has to be careful how they structure their systems & lists.

I think that the thing you point out is how to group several systems that are supposed to bracket others into one idea.  That's an interesting challenge, I'll spend some time on it.

player-03 commented:
> the writer of middleware

Which is also me... for now. But I'm planning ahead.

> For example, if I put a clean-up function in several systems that's supposed to run last (low priority?), after all the systems have done their work, but the user then puts those systems into separate lists, that won't happen anymore.

Edit: oh, I think I get what you're saying. Yes the user can mess things up by creating a separate `SystemList` for everything, but I'm not too worried about that.

(And yeah, low priority runs last.)

onehundredfeet commented:
Cool. All good then.

player-03 committed:
Mention networking instead of singletons.

After thinking it over, multiplayer game developers might consider networking a must-have feature. Those developers would want to know about it, even if the implementation isn't done yet.

Meanwhile, I doubt singletons are anyone's must-have feature.
